### ISSUE TITLE
Line graph wrong when data contains empty values

### ISSUE THREAD

**Issue by fthommen** [Rello/analytics#181]
Consider the following textfile:
```
date;FS1;FS2;FS3
2020-02-03;71,25%;92,45%;null
2020-02-17;71,06%;91,14%;null
2020-02-27;71,80%;91,36%;null
2020-03-02;71,34%;91,56%;null
2020-03-09;71,94%;91,83%;null
2020-03-16;72,68%;92,20%;null
2020-03-23;72,50%;91,97%;null
2020-03-30;73,61%;92,09%;null
2020-04-06;73,86%;83,18%;null
2020-04-15;74,13%;83,23%;null
2020-04-20;74,33%;83,90%;51,35%
2020-04-27;74,88%;84,38%;51,64%
2020-05-11;75,13%;84,39%;51,99%
2020-05-18;77,04%;82,13%;52,31%
2020-05-25;77,11%;82,94%;52,47%
2020-06-15;76,55%;84,89%;53,09%
2020-06-22;76,61%;85,17%;53,30%
2020-06-29;null;85,45%;53,42%
2020-07-06;null;86,84%;90%
2020-07-13;null;87,45%;100%
```

This data is loaded through Advanced configuration -> Dataload with column definitions "FS1,1,2", "FS2,1,3" and "FS3,1,4" respectively and results in the following line graph:

![image](https://user-images.githubusercontent.com/279232/122081186-2ec5c680-cdff-11eb-8304-48ebf6aded59.png)

Problems here are:

  * FS3 data is shifted "up" in time, filling up the empty values.  Eg. the value of 2020-04-20 is printed on 2020-02-03
  * The graph ends at 2020-06-22 instead of 2020-07-13. I.e. it ends when the first empty data for FS1 appears

However the data table does contain all data with the correct dates.

I have also tried with literal "" (empty string) instead of "null" and with just leaving the empty fields completely empty.  It didn't change the outcome.

I am probably doing something wrong here.  But what?

**Comment by fthommen** [comment]
I found that when I display the data as timeline - which it in fact is - the data is displayed with correct start and ends (taking into account missing data in the beginning and the end of the series). But if I add some gap in-between, it is not displayed as gap, but the data is interpolated.  Not sure if this is an other, unrelated problem (I'm happy to open as separate issue for this, just let me know.

This is the updated data table:
```
date;FS1;FS2;FS3
2020-02-03;71,25%;92,45%;null
2020-02-17;71,06%;91,14%;null
2020-02-27;71,80%;91,36%;null
2020-03-02;71,34%;91,56%;null
2020-03-09;71,94%;91,83%;null
2020-03-16;72,68%;92,20%;null
2020-03-23;72,50%;91,97%;null
2020-03-30;73,61%;92,09%;null
2020-04-06;73,86%;83,18%;null
2020-04-15;74,13%;83,23%;null
2020-04-20;74,33%;83,90%;51,35%
2020-04-27;74,88%;84,38%;51,64%
2020-05-11;60,00%;84,39%;51,99%
2020-05-18;null;82,13%;52,31%
2020-05-25;null;82,94%;52,47%
2020-06-15;null;84,89%;53,09%
2020-06-22;76,61%;85,17%;53,30%
2020-06-29;null;85,45%;53,42%
2020-07-06;null;86,84%;90%
2020-07-13;null;87,45%;100%
```

The column definitions have not been changed.

This is how the data is displayed as line chart:
![image](https://user-images.githubusercontent.com/279232/122366817-3058d100-cf5c-11eb-85e0-77bcef945771.png)

And this is how it is displayed as timeline:
![image](https://user-images.githubusercontent.com/279232/122366930-4cf50900-cf5c-11eb-9fe1-5cf02f146010.png)

**Comment by Rello** [comment]
Hello,
this is exactly what is expected. I wanted to test the same with your data but now you came first.

normal line chart is the same as a column chart.
all numbers have to match the x-axis categories. if you put your table into excel, it would do the same.
but here you have one special case: when adding the numbers, "null = not existing" is not exiting and not added to the database.
So FS3s first data record is 51,35% and this will be displayed with the first x-category-value 2020-02-03 because the x-axis is derived from the first dataset and is expected to be the same for all.

long story short: timeline is the way to go because timeline is in fact a chart drawn by x/y coordinates and not by columns. that also explains why a gap is being connected. if it would be a real "0" it would go down.

so pretty much works as designed

**Comment by fthommen** [comment]
Hi,

> normal line chart is the same as a column chart.
> all numbers have to match the x-axis categories. if you put your table into excel, it would do the same.
> but here you have one special case: when adding the numbers, "null = not existing" is not exiting and not added to the database.
> So FS3s first data record is 51,35% and this will be displayed with the first x-category-value 2020-02-03 because the x-axis is derived from the first dataset and is expected to be the same for all.

Hmm.  I probably don't agree on this assessment ;-).  As I understand it, column/line charts are still pairs of x/y coordinates, but not ordered by their value but in the order as they appear in the datafile.  Excel might not be the best comparison here, as Excel does lots of assumptions and has sometimes it's own (irritating) idea of what type your data is.  (most of the data corruption @work results from using Excel :-).

However, this dataset, but with empty fields instead of the "null"s, results in the following (correct) graph in Excel:
![image](https://user-images.githubusercontent.com/279232/122568741-607e9d80-d04a-11eb-93ff-87c86fc652e6.png)

All data series are plotted correctly. The analytics app gives me as many errors as there are empty fields and does not plot the series correctly

When changing e.g. the 1st column  field "2020-02-27" to "2022-12-17" (given, that the fields are formatted as "text" and not "auto" or "date"), the graph doesn't change apart from one label in the x-axis:
![image](https://user-images.githubusercontent.com/279232/122570360-f830bb80-d04b-11eb-9a6d-89d4e114d8fd.png)

Same if I have numbers (field type: number) in the first column (note the 120 which is placed between 5 and 7):
![image](https://user-images.githubusercontent.com/279232/122571223-c8ce7e80-d04c-11eb-96f5-d57dc5a0ecb0.png)


The dataset describes three data series, each with an x and an y coordinate.  In a line/column chart the x coordinate is the position of the data (y value) in the dataset.  I would expect (any) tool, to respect these positions and not shifting (=changing) the x coordinates of any data series. Excel also respects the positions if I create three different datasets and combine them into a single graph.

Do you see that differently?


> long story short: timeline is the way to go because timeline is in fact a chart drawn by x/y coordinates and not by columns. that also explains why a gap is being connected. if it would be a real "0" it would go down.

I am absolutely fine with the timeline, but the timeline doesn't show the gaps as gaps and dataload gives me an error on each missing data.  How to handle the missing data?  "0" is  not a solution, as this marks a value, not a missing data point.

**Comment by Rello** [comment]
Hi,
the issue is, that the datasets are not completely comparable.

A time line will always connect the dots - thats how a timeline will work. due to this, non-related data can be displayed in one chart. imagine your data would have timestamp format because they come from different readings. they would not meet the same millisecond. how should a timeline separate between "no-value" and "time between values"? take 1 measurement per hour, would the seconds between be missing?
what you could try it to overwrite chart options and set the `spanGaps` parameter somehow. it seems to do this, but I don´t know how/if this will work for your chart
https://www.chartjs.org/docs/latest/samples/scales/time-max-span.html

one technical thing: "null" values are never added to the database during the dataloading. this is related to the requirement that a chart needs a value and during the insert of e.g. big batch files, incomplete records should be removed.

**Comment by Himra** [comment]
Hello fthommen,
I had the same problem and solved as follows #161 
maybe this helps

Greeting Himra

**Comment by Rello** [comment]
thank you @Himra for supporting
